lib/compress/zstd_cwksp.h

```diff
@@ -198,29 +212,112 @@ MEM_STATIC size_t ZSTD_cwksp_alloc_size(size_t size) {
+ */
+MEM_STATIC size_t ZSTD_cwksp_slack_space_required(void) {
+    /* For alignment, the wksp will always allocate an additional n_1=[0, 63] bytes
+     * to align the beginning of tables section, as well as another n_2=[0, 63] bytes
+     * to align the beginning of the aligned secion.
+     * 
+     * Then, we also must allocate an additional tableExtraAlignmentBytes=(64 - n_1) +
+     * alignedExtraAlignmentBytes=(64 - n_2) bytes at the end of the workspace to ensure
+     * that we always use an additional 128 bytes in four separate allocations for alignment purposes.
+     */
+    size_t const slackSpace = 4*ZSTD_cwksp_alloc_size(ZSTD_CWKSP_ALIGNMENT_BYTES/2);
+    DEBUGLOG(5, "cwksp: slack space required: %zu", slackSpace);
+    return slackSpace;
+}
+
+/**
+ * Return the number of additional bytes required to align a pointer to the given number of bytes.
+ * alignBytes must be a power of two.
+ */
+MEM_STATIC size_t ZSTD_cwksp_bytes_to_align_ptr(void* ptr, const size_t alignBytes) {
+    size_t const alignBytesMask = alignBytes - 1;
+    size_t const bytes = (alignBytes - ((size_t)ptr & (alignBytesMask))) & alignBytesMask;
+    assert((alignBytes & alignBytesMask) == 0);
+    return bytes;
+}
+
 MEM_STATIC void ZSTD_cwksp_internal_advance_phase(
         ZSTD_cwksp* ws, ZSTD_cwksp_alloc_phase_e phase) {
     assert(phase >= ws->phase);
     if (phase > ws->phase) {
+        /* Going from allocating buffers to allocating objects */
         if (ws->phase < ZSTD_cwksp_alloc_buffers &&
                 phase >= ZSTD_cwksp_alloc_buffers) {
             ws->tableValidEnd = ws->objectEnd;
         }
+
+        /* Going from allocating objects to allocating aligneds */
         if (ws->phase < ZSTD_cwksp_alloc_aligned &&
                 phase >= ZSTD_cwksp_alloc_aligned) {
-            /* If unaligned allocations down from a too-large top have left us
-             * unaligned, we need to realign our alloc ptr. Technically, this
-             * can consume space that is unaccounted for in the neededSpace
-             * calculation. However, I believe this can only happen when the
-             * workspace is too large, and specifically when it is too large
-             * by a larger margin than the space that will be consumed. */
-            /* TODO: cleaner, compiler warning friendly way to do this??? */
-            ws->allocStart = (BYTE*)ws->allocStart - ((size_t)ws->allocStart & (sizeof(U32)-1));
-            if (ws->allocStart < ws->tableValidEnd) {
-                ws->tableValidEnd = ws->allocStart;
+            /* Align the start of the "aligned" to 64 bytes. All aligneds are sized in multiples
+               of 64 bytes, so all subsequent aligneds will naturally be aligned to 64 bytes */
+            void* alloc;
+            size_t bytesToAlign = ZSTD_CWKSP_ALIGNMENT_BYTES - ZSTD_cwksp_bytes_to_align_ptr(ws->allocStart, ZSTD_CWKSP_ALIGNMENT_BYTES);
+            if (bytesToAlign == ZSTD_CWKSP_ALIGNMENT_BYTES) bytesToAlign = 0;
```

Comment:
nit: This would be more simply represented as `(uintptr_t)ws->allocStart & ZSTD_CWSKP_ALIGNMENT_MASK`.

Reply:
Nice! though iirc `uintptr_t` isn't portable enough for the CI tests.